---
title: "tugHall version 3.0: USER-GUIDE"
#authors: "Iurii Nagornov, Nishino Jo and Mamoru Kato"
## date: "`r Sys.Date()`"
output: 
    rmarkdown::pdf_document:
classoption: landscape
linkcolor: Mahogany
urlcolor: blue
---


<!-- 

rmarkdown::pdf_document: 

OR  

rmarkdown::html_vignette:
  vignette: >
    %\VignetteIndexEntry{Vignette Title}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
-->

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  fig.path='./Figures/',
  echo=FALSE, 
  warning=FALSE, 
  message=FALSE,
  collapse = TRUE,
  comment = "#>"
)
options(knitr.table.format = "pipe" )
Sys.setenv("TimeZone" = "Japan")
```

## Requirements for tugHall simulation:

R version **3.6.0** or later

libraries: **stringr, actuar, tidyr**  

**tugHall** is a tool based on the model described in the paper
[Iurii S Nagornov,  Mamoru Kato. tugHall: a simulator of cancer-cell evolution based on the hallmarks of cancer and tumor-related genes. Bioinformatics, V.36, N11, June 2020, pp. 3597–3599](https://doi.org/10.1093/bioinformatics/btaa182). The parameters of the model are described in the Supplementary materials of the paper. 

Note that the program has two different procedures in general: the first is the simulation and the second is the analysis of the simulation results.
Please, pay attention that the requirements for these procedures are **different**.
This User-Guide pertains to the **simulation procedure** alone. 
Please, also note that plots and tables of this document are related to the data files from **/Documentation/Example/** folder. 

<br />

# Table of Contents
1. [Quick start guide](#quick)
2. [Structure of directories](#directories)
3. [Inputs](#inputs)
4. [Outputs](#outputs)
5. [How to run](#run)
6. [Differences with cell-based code and version 2.0](#new)
7. [Differences with clone-based code and version 2.1](#new_cna)

<a name="quick"></a>

# 1. Quick start guide {#quick}

The simplest way to run tugHall:

- Save the **/CNA/** directory to the working folder;
- Run **tugHall_3.0.R**.

The code has its initial input parameters and input files in the **/Input/** folder. After the simulation the user can see results of the simulation (please, see **User-Guide-Analysis_3** for details), which will save to the **/Output/** and **/Figures/** folders. Note that the analysis procedure requires additional libraries and a higher version of R - 3.6.0.

To test code, please, be kind to run tests in the **/Tests/** folder, for details see **/Tests/README.md** file.

<a name="directories"></a>

# 2. Structure of directories {#directories}

### The root directory **/CNA/**:

dir **/Code/** - the folder with a code and a function library.

dir **/Input/** - the folder with the input files.

dir **/Output/** - the folder with the output files.

dir **/Figures/** - the folder with the plot figures.  

dir **/Documentation/** - the folder with documentation and example of a simulation. 

dir **/Tests/** - the folder with tests to check functions of tugHall. 

file **tugHall_3.0.R** - R script to run a simulation and to define the parameters.

file **LICENSE** - the license of the usage of the software based on GNU GENERAL PUBLIC LICENSE version 3.

file **README** - the ReadMe file in markdown format.

file **Sim_monitoring.txt** - the file which is generated during simulation to show the running process.

<br />


### **/Documentation/** directory

file **User-Guide-tugHall_v_3.0** - user guide for a simulation in Rmd or pdf formats.

file **User-Guide-Analysis_v3.0** - user guide for the generation of an analysis and a report  in Rmd or pdf formats.

dir **/Example/** - there are files of a simulation as the data resource for tables of the documentation. 

<br />


### **/Code/** directory:

**tugHall_3.0_functions.R** - the file that contains the functions for the simulation / core of program.

**read_maps.R** - the file to read chromosomal locations got genes of interest from CCDS.current.txt file in the *Input/* folder.

**Functions_clones.R** - the file with the functions for the analysis of results. 

**my_plots.R** - the file with the functions to plot results of a simuation.

<br />


### **/Input/** directory:

**cloneinit.txt** - the file with a list of initial cells with/without destroyed genes. 

**gene_hallmarks.txt** - the file with hallmark variables and weights.

**CCDS.current.txt** - the file with information about chromosomal locations that was getting from [CCDS database](https://ftp.ncbi.nlm.nih.gov/pub/CCDS/current_human/). 

**CF.txt** - the file with the coefficients of Compaction Factors for each hallmark.

**gene_map.txt** - the file with information about chromosomal locations for *genes of interest* only.

**parameters.txt** - the file to read all the parameters which are used in a simulation.

<br />

### **/Output/** directory:

**cloneout.txt** - the file with simulation output.

**geneout.txt** - the file with information about hallmark variables and the weights.

**log.txt** - the file with information about all the parameters. 

**Weights.txt** - the file with information about weights between hallmarks and genes.

**point_mutations.txt** - the file contains information about point mutations in genome of clones.

**CNA_mutations.txt** - the file contains information about copy number alterations in genome of clones.

**gene_map.txt** - the file with information about chromosomal locations for *genes of interest* only.

**order_genes_dysfunction.txt** - the file with data of order of genes' dysfunction.

**VAF.txt**  and **VAF_data.txt** - the files with data related to variant allele frequencies for each mutated site of genes.

<br />

### **/Tests/** directory:

**README.md** - the file with short description how to run tests.

**tests_clones.R** - the file with tests for the functions related to operations with clones.

**tests_cna.R** - the file with tests for the functions related to copy number alterations.

**tests_Model.R** - the file with tests for the main function Model and a simulation.

**/GENE_MAP/**, **/Clones/**, **/CNA/**, **/Model/**, **/Input/** are directories with correct data to compare with the data from running tests.

<br />

### **/Figures/** directory

In the **/Figures/** directory there are figures in \*.pdf format which will appear after the analysis of the simulation results. See **USER-GUIDE-Analysis_3**.

<br />


<a name="inputs"></a>

# 3. Inputs {#inputs}

## Input of hallmark variables and gene weights

The file **CNA/Input/gene_hallmarks.txt** defines the hallmark variables and weights:

```{r, echo=FALSE, results='asis'}
x <- read.csv(file = "../Tests/Input/gene_hallmarks.txt",header = FALSE, sep = "\t" )
x <- data.frame( x[ , 1], x[ , 3], x[ , 2], x[ , 4] )
knitr::kable(x, col.names = c("Genes", "Suppressor or Oncogene", "Hallmark", "Weights"), 
             align = "c", 
             caption = "**Table 1. Input file for genes.** Example of input file for hallmarks and weights in the file **CNA/Tests/Input/gene_hallmarks.txt**."
             )
```

1. **Genes**  - name of gene, e.g., TP53, KRAS. The names must be typed carefully. The program detects all the unique gene names.

2. **Suppressor or oncogene.**  - Distinction of oncogene/suppressor:
- o: oncogene
- s: suppressor
- ?: unknown (will be randomly assigned)
Note that gene malfunction probabilities shown below for "Suppressor" and "Oncogene" are defined separately.

3. **Hallmark** - hallmark name, e.g., "apoptosis". Available names: 
- apoptosis
- immortalization
- growth
- anti-growth
- angiogenesis
- invasion

Note that "growth" and "anti-growth" are related to the single hallmark "growth/anti-growth".
Note that "invasion" is related to "invasion/metastasis" hallmark.


4. **Weights**  - Hallmark weights for genes, e.g., 0.333 and 0.5. For each hallmark, the program checks the summation of all the weights. If it is not equal to 1, then the program normalizes it to reach unity. Note that, if the gene belongs to more than one hallmark type, it must be separated into separate lines.

---


After that, the program defines all the weights. **Unspecified  weights** are set to 0. Program performs normalization so that the sum of all weights should be equal to 1 for each column (see next table). The **CNA/Output/Weights.txt** file saves these final input weights for the simulation.

```{r, echo=FALSE, results='asis'}
x <- read.csv(file = "../Documentation/Example/Weights.txt", header = TRUE, sep = "\t", nrows = 10)
knitr::kable(x,  col.names = c("Genes", "Apoptosis, $H_a$", "Angiogenesis, $H_b$", "Growth / Anti-growth, $H_d$", "Immortalization, $H_i$", 
                               "Invasion / Metastasis, $H_{im}$"), align = "c", caption = "**Table 2. Weights for hallmarks.** Example of weights for hallmarks and genes from _**CNA/Documentation/Example/Weights.txt**_ file. Unspecified values equal 0."  )
```

1. **Genes** - name of genes.

2. **Apoptosis, $H_a$** - weights of hallmark "Apoptosis".

3. **Angiogenesis, $H_b$** - weights of hallmark "Angiogenesis".

4. **Growth / Anti-growth, $H_d$** - weights of hallmark "Growth / Anti-growth".

5. **Immortalization, $H_i$** - weights of hallmark "Immortalization".

6. **Invasion / Metastasis, $H_{im}$** - weights of hallmark "Invasion / Metastasis".

---

## Input the probabilities

The input of the probabilities used in the model is possible in the code for parameter value settings, see function **define_paramaters()** in the file **"tugHall_3.0.R"**:

|  Probability variable and value | Description  |  Units  |
|:---|:---|:---|
| **E0 = 2E-4**    | Parameter $E_0$ related to environmental resource limitation |  $^*$ |
| **F0 = 1E0**     | Parameter $F_0$ related angiogenesis  |  $^*$ |
| **m =  1E-6**    | Point mutation probability  $m'$ | per cell's division per base pair |
| **uo = 0.5**     | Gene malfunction probability by point mutation for oncogene $u_o$ | per mutation  |
| **us = 0.5**     | Gene malfunction probability by point mutation for suppressor $u_s$ |  per mutation |
| **s =  10**      | Parameter in the sigmoid function $s$ |  $^*$ |
| **k =  0.1**     | Environmental death probability $k'$ |  per time-step  |
| **m_dup      =  0.01**     | CNA duplication probability $m_{dup}$ | per cell's division  |
| **m_del      =  0.01**     | CNA deletion probability $m_{del}$ |  per cell's division  |
| **lambda_dup =  7000**        | CNA duplication average length $\lambda_{dup}$  | the geometrical distribution for the length  |
| **lambda_del =  5000**        | CNA deletion average length $\lambda_{del}$  |  the geometrical distribution for the length |
| **uo,dup = 0.8**     | Gene malfunction probability by CNA duplication for oncogene $u_{o,dup}$ |  per mutation |
| **us,dup = 0**     | Gene malfunction probability by CNA duplication for suppressor, $u_{s,dup}$. Currently, 0 is assumed. | per mutation  |
| **uo,del = 0**     | Gene malfunction probability by CNA deletion for oncogene $u_{o,del}$. Currently, 0 is assumed. | per mutation  |
| **us,del = 0.8**     | Gene malfunction probability by CNA deletion for suppressor, $u_{s,del}$.  | per mutation  |
| **d0 = 0.35**     |  Initial division rate  | per time-step  | 
| **censore_n = 30000 **     | Max cell number where the program forcibly stops  | number of cells |
| **censore_t = 200**        | Max time where the program forcibly stops  | in time-steps |
| **Compaction_factor	= TRUE**   | Indicator about an usage of compaction factor | Logical |
| **model_name	= 'proportional_metastatic'**  | Model definition, it can be 'proportional_metastatic' or 'threshold_metastatic' or 'simplified' |  string/character variable |
| **time_stop	= 120**  | Max time of running after that the program forcibly stops| in seconds |
| **n_repeat	= 1**  | Max number of repetition of the program until the NON-ZERO output will be getting | must be integer |
| **monitor	= TRUE**| The indicator to safe or not to the monitoring file *Sim_monitoring.txt* during a simulation | logical |
| <img width=160/> | <img width=340/> |  <img width=200/> |

$^*$ [see Suplementary materials in Bioinformatics,V.36,N11,2020,p.3597](https://doi.org/10.1093/bioinformatics/btaa182)

User can also define input parameters from the file **parameters.txt** and print all the parameters like:

    define_paramaters( read_fl = TRUE , file_name = './Input/parameters.txt' )
    print_parameters()

---

## Compaction factor

If the model 'proportional_metastatic' is used then an user should to define compaction factors which reduce hallmark values like:

    define_compaction_factor( read_fl = TRUE , file_name = './Input/CF.txt' )
where file **'CNA/Input/CF.txt'** contents data of compaction factors:

|  Hallmark's name | Factor  |
|:---|:---|
| apoptosis | 0.9 |
| growth | 	0.85 |
| immortalization | 	0.79 |
| angiogenesis | 	0.82 |
| invasion | 0.97 |
| <img width=160/> | <img width=140/> |  
---

## Filename input

Also in the code **"tugHall_3.3.R"** user should define names of input and output files using function **define_files_names()** before a simulation:

|  Variables and file names  | Description  |
|:---|:---|
| **genefile = 'gene_hallmarks.txt'**     | File with information about gene-hallmarks weights  |
| **mapfile = 'gene_map.txt'**     | File with information about genes' map |
| **clonefile = 'cloneinit.txt'**      | Initial Cells  |
| **geneoutfile = 'geneout.txt'**    | Gene Out file with hallmarks  |
| **cloneoutfile = 'cloneout.txt'**   | Output information of simulation  |
| **logoutfile =  'log.txt'**       | Log file to save the input information of simulation  |
| <img width=200/> | <img width=350/> |


---


## Input of the initial clones

**Please, pay attention, it works for driver point mutation only.**

The initial states of cells are defined in **"CNA/Input/cloneinit.txt"** file:

|  Clone ID  | List of malfunctioned genes  | Number of cells |
|:---|:---|:---|
| 1 | ""  | 1000 |
| 2    |  "APC" | 10 |
| 3   | "APC, KRAS"  | 100 |
| 4   | "KRAS"  | 1 |
| 5   | "TP53, KRAS"  | 1 |
| ...   | ...  | 100 |
| 1000        | ""  | 10 |
| <img width=50/> | <img width=150/> | <img width=150/> |

1. **Clone ID**  - ID of clone, e.g., 1, 324.

2. **List of malfunctioned genes** - list of malfunctioned genes for each clone, e.g. "“,”KRAS, APC“. The values are comma separated. The double quotes ("") without gene names indicate a clone without malfunctioned genes.

3. **Number of cells**  - number of cells in each clone, e.g., 1, 1000.


## Input of the genes' maps

This new version of **tugHall** allows to calculate CNAs in the genome. The breakpoints of CNAs may fall on genic regions consisting of exons and introns. That's why it's needed to enter information about gene's map.
In the **/Input/** directory you can find **CCDS.current.txt**, which was getting from [CCDS database](https://ftp.ncbi.nlm.nih.gov/pub/CCDS/current_human/) at the National Center for Biotechnology Information and has information about genes. 
At the beginning of simulation, the program reads this file and extracts genes’ map using function **define_gene_location()**, which is put into 
 **"CNA/Input/gene_map.txt"**. For example, the map is shown as follow:

```{r, echo=FALSE, results='asis'}
x <- read.table( file = "../Input/gene_map.txt",header = TRUE, sep = "\t",  nrows = 10)
knitr::kable(x, align = "c", caption = "")
```

1. **Chr**  - Name of the chromosome, e.g., 1, 12, X, Y.

2. **CCDS_ID** - ID of the gene in the [CCDS database](https://www.ncbi.nlm.nih.gov/CCDS/).

3. **Gene**  -  the name of the gene.

3. **Start**  -  the start position of each exon of the gene.

4. **End**  -   the final position of each exon of the gene.

5. **Len** - the length of gene's location *Len = End - Start + 1*

---


\newpage

<a name="outputs"></a>

# 4. Outputs {#outputs}

The output data consists of several files after the simulation. 

## "log.txt" file
The file **"log.txt"** contains information about probabilities and file names. These variables are explained in the ["Inputs"](#inputs). 

```{r, echo=FALSE, results='asis'}
x <- read.csv(file = "../Documentation/Example/log.txt",header = FALSE, sep = "\t", col.names = c("Variable","Value"), stringsAsFactors = FALSE )
x[is.na(x)] <- ""
x[ c(6,7) , 1 ] =  c( 'E0' , 'F0' )
knitr::kable(x, align = "c", caption = "**The log.txt file.** Example of log.txt file.")
```


## "geneout.txt" file
The file **"geneout.txt"** contains input information about the weights that connect the hallmarks and genes, which are defined by the user. These variables also are explained in the ["Inputs"](#inputs). 

```{r, echo=FALSE, results='asis'}
x <- read.csv(file = "../Documentation/Example/geneout.txt",header = FALSE, sep = "\t", col.names = c("Gene_name","Hallmark_name", "Weight", "Suppressor_or_oncogene"), stringsAsFactors = FALSE )
x[is.na(x)] <- ""
knitr::kable(x, align = "c", caption = "**The geneout.txt file.** Given below is an example of the geneout.txt file.")
```

\newpage 

## "cloneout.txt" file

The file **"cloneout.txt"** contains the results of the simulation and includes the evolution data: all the output data for each clone at each time step (only the first 10 lines are presented): 

```{r, echo=FALSE, results='asis'}
x <- read.csv(file = "../Documentation/Example/cloneout.txt",header = TRUE, sep = "\t", nrows = 10, stringsAsFactors = FALSE )
x <- x[,c(1:13)] 
# library(stringr)
# library(tidyr)
x[is.na(x)] <- ""
library(stringr)
x[] <- lapply(x, str_trunc, 6, ellipsis = "")
names( x )[ 6 ] = 'Birth_time'
knitr::kable(x, align = "c", caption = "**The Output data from cloneout.txt file.** Example of output data for all clones. The names of columns are related to the description in the Tables 1,2 and *USER-GUIDE-Analysis_3*'s figures. Columns are from 1 to 13.")
```

1. **Time** - the time step, e.g., 1, 50.
2. **N\_cells** - the number of cells in this clone, e.g. 1000, 2.
3. **AvgOrIndx** - "avg" or "index": "avg" is for a line with averaged values across different (index) lines at the same time step; "index" shows the cell's index at the current time step,  e.g., avg, 4,7.
4. **ID** - the unique ID of clone, e.g., 1, 50.
5. **Parent_ID** - the parent index,  e.g., 0, 45.
6. **Birth_time** - the time step of the clone's birth,  e.g., 0, 5.
7. **c** - the counter of cell divisions for the clone, it equals average counter across all the cells in the clone.
8. **d**  - the probability of division for the cell, e.g., 0.1, 0.8 [per time-step].
9. **i**  - the probability of immortalization for the cell, e.g., 0.1, 0.8 [per time-step].
10. **im**   - the probability of invasion/metastasis for the cell, e.g., 0.1, 0.8 [per time-step].
11. **a**   - the probability of apoptosis for the cell, e.g., 0.1, 0.8 [per time-step].
12. **k**   - the probability of death due to the environment, e.g., 0.1, 0.8 [per time-step].
13. **E**  - the E coefficient for the function of the division probability, e.g., 10^4, 10^5.

\newpage

```{r, echo=FALSE, results='asis'}
x <- read.csv(file = "../Documentation/Example/cloneout.txt",header = TRUE, sep = "\t", nrows = 18, stringsAsFactors = FALSE )
x[is.na(x)] <- ""
x <- x[,c(1,3,14:24)] 
library(stringr)
x[,c(1:11,13)] <- lapply(x[,c(1:11,13)], str_trunc, 7, ellipsis = "")
#names(x)[10] = 'driver_genes'
knitr::kable(x, align = "c", caption = "**cloneout.txt** Columns are from 14 to 24.")
```

14. **N_normal** - the number of normal cells at this time step, e.g., 134, 5432.
15. **Nmax** - the theoretically maximal number of primary tumor cells, e.g., 10000, 5000.
16. **N_primary** - the number of primary tumor cells at this time step, e.g., 134, 5432.
17. **N_metastatic** - the number of metastatic cells at this time step, e.g., 16, 15439.
18. **Ha** - the value of the hallmark "Apoptosis" for the cell, e.g., 0.1, 0.4444.
19. **Him** - the value of the hallmark "Invasion / Metastasis" for the cell, e.g., 0.1, 0.4444.
20. **Hi** - the value of the hallmark "Immortalization" for the cell, e.g., 0.1, 0.4444.
21. **Hd** - the value of the hallmark "Growth / Anti-growth" for the cell, e.g., 0.1, 0.4444 .
22. **Hb** - the value of the hallmark "Angiogenesis" for the cell, e.g., 0.1, 0.4444 .
23. **type** - the type of the cell: 'normal' or 'primary' or 'metastatic'.
24. **mut_den** - the density of mutations for the cell, it equals to ratio a number of mutated driver genes to a number of all the genes, e.g., 0, 0.32.

\newpage
```{r, echo=FALSE, results='asis'}
x <- read.csv(file = "../Documentation/Example/cloneout.txt",header = TRUE, sep = "\t", nrows = 18, stringsAsFactors = FALSE)
x[is.na(x)] <- ""
x <- x[,c(1,3,25:28)] # ncol(x)] 
knitr::kable(x, align = "c", caption = "**cloneout.txt** Columns are from 25 to 28.")
```
25. **driver_genes** - the binary numbers indicate the driver mutation at the gene related to order of genes in onco as well as order of the next columns with genes' names, e.g., '1 0 0 0' means that the first gene has a driver mutation and other genes have no.
26. **passenger_genes** - the binary numbers indicate the passenger mutation at the gene related to order of genes in onco as well as order of the next columns with genes' names, e.g., '0 0 1 0' means that the third gene has a passanger mutation and other genes have no.
27. **PointMut_ID** - the index of data row for point mutation data frame saved at the end of simulation in the file **Point_mutations.txt**, e.g., 23, 32.
28. **CNA_ID** - the index of data row for CNA data frame saved at the end of simulation in the file **CNA.txt**, e.g., 44, 21.

There are two columns (25th and 26th) with the indexes of point mutations and CNAs. 
Each index corresponds to index in the related data frames for point mutations and for CNAs represented in additional tables below.

\newpage
```{r, echo=FALSE, results='asis'}
x <- read.csv(file = "../Documentation/Example/cloneout.txt",header = TRUE, sep = "\t", nrows = 18, stringsAsFactors = FALSE)
x[is.na(x)] <- ""
x <- x[,c(1,3,29:32)] # ncol(x)] 
knitr::kable(x, align = "c", caption = "**cloneout.txt** Columns are from 29 to 32.")
```

29-32. **Chr1_CDS_(gene's name)**, for example **Chr1_CDS_APC** - the length of CDS for each gene in the order of names of genes for FIRST parental chromosome of a clone. The CDS length of genes for second parental chromosome can be different in principle. The point mutation is proportional to **CDS_(gene's name)** od each chromosome. 


\newpage
```{r, echo=FALSE, results='asis'}
x <- read.csv(file = "../Documentation/Example/cloneout.txt",header = TRUE, sep = "\t", nrows = 18, stringsAsFactors = FALSE )
x[is.na(x)] <- ''
x <- x[, c( 1,3, 33:36 ) ] 
library(stringr)
x[] <- lapply(x, str_trunc, 8, ellipsis = "")
knitr::kable(x, align = "c", caption = "**cloneout.txt** Columns are from 33 to 36.")
```

33-36. **Chr1_Len_(gene's name)**, for example **Chr1_Len_APC** - the length of gene in the order of names of genes for ONLY FIRST chromosome of a clone. The length of genes for second chromosome can be different in principle. The CNA mutation is proportional to **Len_(gene's name)**. 


\newpage
```{r, echo=FALSE, results='asis'}
x <- read.csv(file = "../Documentation/Example/cloneout.txt",header = TRUE, sep = "\t", nrows = 18, stringsAsFactors = FALSE )
x[is.na(x)] <- ''
x <- x[, c( 1,3, 37:40 ) ] 
library(stringr)
x[] <- lapply(x, str_trunc, 8, ellipsis = "")
knitr::kable(x, align = "c", caption = "**cloneout.txt** Columns are from 37 to 40.")
```

37. **Chr1_p0** - the probability that during a trial, a cell of the clone has **NO** mutation [per time-step]. Applied to all cells in the clone.

38. **Chr1_prob_point_mut** - the **conditional** probability that if cell will have a mutation it should be a **point mutation**. 

39. **Chr1_prob_del** - the **conditional** probability that if cell will have a mutation it should be a **deletion**. 

40. **Chr1_prob_dup** - the **conditional** probability that if cell will have a mutation it should be a **duplication**. 


Last columns (from 41 to the end) show the data related to the **SECOND** parental chromosome with the same set of data as for the first parental chromosome.

\newpage
```{r, echo=FALSE, results='asis'}
x <- read.csv(file = "../Documentation/Example/point_mutations.txt",header = TRUE, sep = "\t", stringsAsFactors = FALSE)
# x[is.na(x)] <- ""
x = x[ 1 : 20, ]
knitr::kable(x, align = "c", caption = "Point mutation data frame which will be saved to the file **Point_mutations.txt** at the end of simulation.")
```


1. **PointMut_ID** - ID of point mutation, first ID is related to variant allele 'B' and same *second* ID - to the original allele A. 
2. **Parental_1or2** - indicates either of the two parental chromosomes. 
3. **Chr** - name of a chromosome.
4. **Ref_pos** - the reference position of an allele. The reference position is on the coordinate system of the human reference genome. 
5. **Phys_pos** - the physical position of an allele. The physical length of a (parental) chromosome is extended or shrunk by CNA duplications or deletions, respectively. When a duplication happens, the reference position is divided into two or more physical positions, which are represented by multiple elements in a vector. When a deletion happens and the allele is lost, the lost is represented by “-” on the coordinate system of physical positions. 
6. **Delta** - difference between the reference and physical positions.
7. **Copy_number** - the copy number of an allele. 
8. **Gene_name** - the name of a gene.
9. **MalfunctionedByPointMut** - logical indicator of whether or not the gene is malfunctioned by the point mutation. 
10. **mut_order** - indicator of mutation order in the simulation, it's used to detect order of mutations in the clone at each chromosome.



```{r, echo=FALSE, results='asis'}
x <- read.csv(file = "../Output/CNA_mutations.txt",header = TRUE, sep = "\t", stringsAsFactors = FALSE)
# x[is.na(x)] <- ""
x = x[ 1 : 20, ]
knitr::kable(x, align = "c", caption = "**CNA mutation data frame which will be saved to the file **CNA.txt** at the end of simulation.")
```

1. **CNA_ID** - ID of CNA. 
2. **Parental_1or2** - indicates either of the two parental chromosomes. 
3. **DuplicationOrDeletion** - indicator of duplication or deletion for CNA.
4. **Chr** - name of a chromosome.
5. **Reference_start** - the reference position of the CNA start.
6. **Reference_end** - the reference position of the CNA end.
7. **Gene_name** - the name(s) of a gene(s).
8. **MalfunctionedByCNA** - logical indicator of whether or not the gene(s) is malfunctioned by the CNA. 
9. **mut_order** - indicator of mutation order in the simulation, it's used to detect order of mutations in the clone at each chromosome.

---

\newpage



```{r, echo=FALSE, results='asis'}
x <- read.csv(file = "../Sim_monitoring.txt",header = TRUE, sep = "\t", stringsAsFactors = FALSE)
x = x[ 1 : 20, ]
knitr::kable(x, align = "c", caption = "**Sim_monitoring.txt** file with data getting during a simulation.")
```

1. **Time** - time step. 
2. **N_clones** - total number of clone at a time step. 
3. **N_normal** - total number of normal cells at a time step. 
4. **N_primary** - total number of primary tumor cells at a time step. 
5. **N_metastatic** - total number of metastatic cells at a time step. 
6. **N_point_mutations** - total number of unique point mutations at a time step. 
7. **N_duplications** - total number of unique duplications  at a time step. 
8. **N_deletions** - total number of unique deletions at a time step. 
							
---

\newpage

<a name="run"></a>

# 5. How to run {#run}

In order to make the simulation, please follow the procedure:

1. Copy **/CNA/** directory into the working directory.

2. CD to the **/CNA/** directory. 

3. Run the **tugHall_3.0.R** file, using the command line like

`R --vanilla < tugHall_3_0.R`

or using the line by line procedure in **R Studio** in the **tugHall_3.0.R** file. 

4. To obtain analysis reports of the simulation, please refer to **User-Guide-Analysis_v3.0.RMD**. 
In **User-Guide-Analysis_v3.0.RMD** and **User-Guide-tugHall_v3.0.RMD**, commands are embedded to include files under **/Documentation/Example/**. 


<a name="new"></a>

# 6. Differences with cell-based code and version 2.0 {#new}

## 6.1. Reason to develop clone-based code

* Clone-based code was designed to accelerate calculation and increase number of tumor cell. 
Advantage of clone-based algorithm is making trial for all cells at 1 clone with one application of **trial()** function. In cell-based algorithm **trial()** applies to each cell. But if number of cells equal number of clones, then speed up is 1. That's why  clone-based code works faster for any cases. 

* Another reason is a case, when we need to simulate huge number of cells like $10^7$ or $10^9$, but mutation rate is very low. Cell-based algorithm takes a huge computational cost, and vice versa clone-based algorithm will work very fast, if mutated cells will appear slowly.

## 6.2. Usage of _trial()_ function

* In **trial()** function program applies several trials like environmental death, apoptosis death, division process, etc. We changed the trials with probability $p$ (for some death process) for each cell in the clone with for 1 trial with procedure:

$N_{cells} = N_{cells} - Binom(p,N_{cells})$,

where $Binom(p,N_{cells})$ is random number from the binomial distribution with probability $p$, $N_{cells}$ is a number of cells in a clone. Probability $p$ is one of probabilities of death processes, for example, for apoptosis death $p = a'$ or for environment death $p = k$ etc. 

* For cell division with probability $d'$ the new number of cells will be:

$N_{cells} = N_{cells} + Binom(d',N_{cells})$


* Check at the end of **trial()** function: if $N_{cells} = 0$, then the clone has died.

## 6.3. Usage of mutation function

* In mutation function we have changed the mutation to birth of a new clone (one mutation is a birth of one clone):

$N_{new\_clones} = Binom(m,N_{new\_cells})$,

$N_{new\_cells} = Binom(d',N_{cells})$.

* Passenger or Driver mutations do not matter for new clone's generation. Only during analysis, we will distinguish Passengers or Drivers clones.


## 6.4. Average function

* The average values $\overline{x}$ of probabilities or hallmarks are found by summation on the $x_i$ with multiplication by cells number $N_{cells,i}$ of this clone:

$\overline{x} = \sum_i x_i \times w_i$,

where $w_i = \frac{N_{cells,i} }{ N_{cells,tot}}$ is $i$th clone’s occupancy in
whole cell population $N_{cells,tot} = \sum_i N_{cells,i}$, 
$x_i$  is the value for $i$th clone, 
summation applies for all clones $i = 1 .. N_{clones}$.

* For this purpose, we added the calculation of cells number (primary and metastasis) before average and hallmarks update.


## 6.5. Difference with version 2.0

In the current version we use library *actuar* to make non-zero-binom calculation faster, and 
we use the approximation for big numbers of cells in **trial()** function, because **rbinom()** function in R has restriction for big numbers like $n \times p > 10^{12}$. 


<a name="new_cna"></a>

# 7. Differences with clone-based code and version 2.1 {#new_cna}

## 7.1. Reason to develop CNA-based code

New version of tugHall with copy number alteration (CNA) was designed for correct calculation of VAF influenced by CNA and tumor purity. It's expected that this design should improve comparison between observation VAF $\in [0;1]$ and calculated VAF. The previous versions of tugHall have VAF in the range $[0;0.5]$
because of the neglect of CNA and tumor purity.


## 7.2. Changing the formula of the cell division coefficient

In the cell division process, the logistic growth applies to primary-tumor cells and normal cells, where normal cells are cells without any driver mutations. Meanwhile, $N_p$ of the friction term in the logistic equation is the number of primary-tumor cells. $N_p$, $N_m$, and $N_n$ are the numbers of primary-tumor, metastatic-tumor, and normal cells. So, the division coefficient now is calculated by next formula:

$$
\large d'=
\begin{cases}
(d_0 + H_d) ( 1 - E' \times N_p) , \qquad \text{when logistic growth,} \\
d_0 + H_d, \qquad \qquad \qquad\text{when exponential growth}    \\
\end{cases}
$$
where $H_d$ is division hallmark, 
$d_0$ is initial division coefficient, 
$E' = \frac{E_0}{ (1+F_0 \times H_b) }$,
$H_b$ is angiogenesis hallmark, $F_0$ is a friction coefficient.


## 7.3. Calculation of point and CNA mutations

Probabilities of CNA mutations are calculated in the same way as point mutations:

* $m_{point} = m_0 \times l_{CDS}$ - for point mutation of a gene, where $l_{CDS}$ is the length of all exons of a gene ( *CDS_(gene's name)* is denoted in the table above ) and $m_0$ is a constant per base pairs per cell's division defined by users;


* $m_{0,dup}$ and $m_{0,del}$, or we collectively call $m_{0,CNA}$, indicate the first breakpoint event of a CNA and it is a constant per base pairs per cell's division defined by users. 
$m_{CNA} = m_{0,CNA} \times l_{genes}$, where $l_{genes}$ is the total region size of all genes of interest which consists of exons as well as introns ( *Len_(gene's name)* is denoted in the table above ). 


* a length of CNA is calculated using geometrical distribution:
    $l_{CNA} = rgeom(1,1 / \lambda_{CNA})$+1, where $\lambda_{CNA}$ is average base-pair length defined by users ($\lambda_{CNA}$ is either $\lambda_{dup}$ or $\lambda_{del}$).

* probability of malfunctioning a gene $u = u_{s,CNA}$ for suppressor and 
$u = u_{o,CNA}$ for oncogene.

So, the algorithm of CNA is as follow:

    if ( runif(1) < m_dup + m_del ) then 'Generate CNA':
      - define which event should occur - duplication or deletion using ratio m_dup/m_del like: 
          event  <-  sample(c('dup', 'del'), 1, prob = c( m_dup, m_del )/sum(m_dup, m_del) )

      - find randomly first position within the regions of genes of interest;
      - find the length of CNA from geometrical distribution
      - define with probability 0.5 is it the parental chromosome 1 or 2;
      - define the list of genes in CNA;
      - define with probabilty u = {u_o or u_s} is it malfunction for each gene;
      - check overlap of position for other mutations (and if it's necessesary change their positions).


The calculation of probabilities and hallmarks variables is not changed. 

At the end of a simulation the VAF frequencies are calculated in accordance with formulation:

$$
\LARGE VAF^i = 
\frac{(1-\rho)n^i_{B,N}+ \sum_{s=1}^{\#sp} \tau_s n^i_{B,S}}{(1-\rho)(n^i_{A,N} + n^i_{B,N}) + \sum_{s=1}^{\#sp} \tau_s (n^i_{A,S} + n^i_{B,S}) },
$$

where:

$i$ is position (site) index,

$s$ is subpopulation (clone's) index,

$\tau$ is subpopulation (clone's) fraction, 

$\rho$ is tumor purity: $\rho = \sum_{s=1}^{\#sp} \tau_s$,

$n$ is copy number,

A denotes an original allele A, B - variant B, N - normal, S - tumor.

In usual application we used for normal cells $n^i_{A,N} = 2$ and $n^i_{B,N} = 0$, so VAF can be calculated as follow:

$$
\LARGE VAF^i = 
\frac{\sum_{s=1}^{\#sp} \tau_s n^i_{B,S}}{2(1-\rho) + \sum_{s=1}^{\#sp} \tau_s (n^i_{A,S} + n^i_{B,S}) }
$$



In tugHall, under the given admixture rate of intact normal cells $\rho_N$, values of VAF are calculated using next formulation:

$$
\LARGE VAF^i = 
\frac{(1-\rho_N) \sum_{s=1}^{\#sp} \lambda_s n^i_{B,S}}{2\rho_N + (1-\rho_N) \sum_{s=1}^{\#sp} \lambda_s (n^i_{A,S} + n^i_{B,S}) },
$$

where $\lambda_s = (N^s_T+N^s_{SN})/(N_T+N_{SN})$, ${N_T}$ and $N_{SN}$ are total number of primary tumor and speckled normal cells respectively at the last time step of a simulation, index $s$ related to cells with mutated site $s$. Observation subpopulation (clone's) fraction $\tau_s$ and simulation one $\lambda_s$ are proportional with $\tau_s = (1-\rho_N) \lambda_s$. Please, pay attention that for metastatic cells, we put $\rho_N = 0$.




